design-proposals/backend-storage-clone-on-migration/backend-storage-clone-on-migration.md

```diff
@@ -0,0 +1,105 @@
+
+We have opened RFEs against libvirt for [TPM](https://issues.redhat.com/browse/RHEL-54325) and [EFI](https://issues.redhat.com/browse/RHEL-54326), in case we decide to pursue to this approach in the future.
+
+### Partitioned block support
+
+[A pull request](https://github.com/kubevirt/kubevirt/pull/10346) (updated [here](https://github.com/kubevirt/kubevirt/pull/12337)) explored a not-safely-migratable implementation which partitions the block storage as ext3.  
+Migratability actually appears functional but should be considered unsafe, since 2 pods (source and target virt-launchers) mount the same ext3 filesystem concurrently.  
+Replacing ext3 with `gfs2` could be something to consider to solve this problem.
+
+## Motivation/Goals
+- Users want to be able to use any storage class for backend-storage
+- Users want all VMs with persistent features to be potentially migratable
+
+## Definition of Users
+Any VM owner that wants to use persistent features, such as TPM and/or EFI
+
+## User Stories
+As a user, I want to be able to seamlessly enable persistent features while preserving the migratability of my VMs.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+- When creating a new VM with persistent features, we use either (first match wins):
+  - The storage class specified in the KubeVirt Custom Resource
+  - The storage class marked as the default for virtualization purposes unless it has a StorageProfile that shows only block is supported
+  - The storage class marked as the default for the cluster
+- All new persistent storage PVCs will be created with the name `persistent-state-for-<vm_name>-<random_string>` and the new annotation `persistent-state-for-<vm_name>=true`
+- When starting an existing VM with persistent features, we will look for any PVC with the annotation, or any PVC with the legacy name, for backwards compatibility
+- The volume mode will be Filesystem
+- The access mode will be RWO, unless a StorageProfile exists for the Storage Class and shows only RWX is supported
+  - Note: the CR field documentation will be adjusted to reflect that RWX is no longer needed
+- When a migration is created:
+  - We create a new empty PVC for the target, with the name `persistent-state-for-<vm_name>-<random_string>` and no annotation
+  - In the migration object status, we store the names of both the source and target PVCs (see API section below)
+  - If a migration succeeds, we set the annotation `persistent-state-for-<vm_name>=true` on the new (target) PVC and delete the old (source) PVC, using the source PVC name from the migration object
+  - If a migration fails, we delete the target backend-storage PVC that was just created, using the target PVC name from the migration object
+  - In the unlikely event that a VM shuts down during a migration, the migration will be considered failed and the target PVC will get discarded
```

Comment:
What if right after migration ends successfully in the libvirt level, the VM makes a change to the TPM and shuts down? Discarding the target PVC would make the data lost.

Reply:
Yes, see https://github.com/kubevirt/community/pull/314#discussion_r1791924517

Reply:
This is not a KubeVirt problem, same thing would happen on bare-metal... If one doesn't find the risks of post-copy acceptable, they should just not use it.
What we can and should do is document those risks thoroughly.

Reply:
> What if right after migration ends successfully in the libvirt level, the VM makes a change to the TPM and shuts down? Discarding the target PVC would make the data lost.

Sorry I misread your question. If a libvirt migration ends successfully then the KubeVirt migration will also end successfully and all will be fine.
I read something like "What if right after migration handoff [...]", my bad.

Reply:
Hi @dankenigsberg can you clarify what user scenario you're focused on/concerned about? In all cases, a machine only modifies the registers of its TPM during initial power on. The protected contents are only changed if the user changes their LUKS settings (or windows equivalent). This is not the same as boot order type changes. this is setting the initial password for your encrypted volume. This is not a sensible thing to do immediately before shutdown.

Reply:
I think it is a valid concern.

> > What if right after migration ends successfully in the libvirt level, the VM makes a change to the TPM and shuts down? Discarding the target PVC would make the data lost.
> 
> Sorry I misread your question. If a libvirt migration ends successfully then the KubeVirt migration will also end successfully and all will be fine. I read something like "What if right after migration handoff [...]", my bad.

KubeVirt finalizes the migration in the source pod. It relies on getJobStats which is being retrieved every 400ms.
There is a period of time (less than 400ms) during which the migration can potentially be finalized but the metadata won't be updated.
If something happens to the source pod at this time and the target manages to write to the new volume, we will be in trouble.

Reply:
> If something happens to the source pod at this time and the target manages to write to the new volume, we will be in trouble.

Could you please detail this scenario a bit more? Are we still talking about a failed post-copy migration? Also what do you mean by "something happens to the source pod"?
Thanks!

Reply:
> "something happens to the source pod"

e.g the node running it can crash.

Also, the VM can finish cleanly or crash during the window between libvirt's end of migration and etcd being aware that migration ended successfully. The guest may write everything to its storage and tpm, but we end up deleting the tpm.

Reply:
> > If something happens to the source pod at this time and the target manages to write to the new volume, we will be in trouble.
> 
> Could you please detail this scenario a bit more? Are we still talking about a failed post-copy migration? Also what do you mean by "something happens to the source pod"? Thanks!

Sure. Consider a regular pre-copy migration.
libvirt completed the migration and unpaused the target.
The migration results to kubevirt objects take some time to propagate (speaking in general).
If the source pod is terminated (due to node crash, OOM, kubelet killing it) and the target writes the TPM state to the new volume during this window the target volume will be lost.

Reply:
I see. To repeat back and summarize, if all the following is true we have a problem:

- We're inside the, at most 400ms, window when the libvirt migration is finished but KubeVirt doesn't know yet
- The source pod/node crashes
- The VM writes to the TPM

Then the VMIM will fail and that TPM write will be dropped.
Honestly, to me, this scenario feels quite statistically impossible...
If we want to address this, do you have a suggestion how?
I brainstormed this with @stu-gott and this is the only thing we could come up with:

- Libvirt+swtpm implement a state lock that's automatically enabled at the end of a migration
- Once KubeVirt sees that the libvirt migration succeeded, it instructs libvirt to release that lock

If that makes sense to you we can discuss it with the libvirt team.
However, would you agree to do that as a follow-up effort?

Thanks!

Reply:
The 400ms is only the time it takes for the migration monitor to update the domain metadata.
Then there is also the time required to propagate the migration status to other kubevirt components/objects.
Overall, it's possible that this can be more than just a statistically insignificant issue.

I didn't have much time to think this through.
However, If we are going to continue with this effort we need to come up with a mechanism that will allow us to choose the right volume on the next VM startup.
I don't think that locking is enough - I don't see how it solves the issue. It doesn't look like there is a competing process that can write into the source volume. 

Instead of discarding the target volume for unfinalized migrations, we should keep it around.
Then, on the next VM start up we can select the volume with the latest time stamp.

Reply:
> Instead of discarding the target volume for unfinalized migrations, we should keep it around.
> Then, on the next VM start up we can select the volume with the latest time stamp.

We can easily change the code to keep target volumes around for migrations that didn't succeed.
However, that timestamp thing doesn't make sense to me. We have no idea what the state of the target PVC is as long as we don't see the migration succeed. It could very well be empty or partially populated. What we can do is keep it around and clean it up next time we migrate the VM? That way it's there for recovery purposes in case the highly unlikely scenario above occurs.

Reply:
> > Instead of discarding the target volume for unfinalized migrations, we should keep it around.
> > Then, on the next VM start up we can select the volume with the latest time stamp.
> 
> We can easily change the code to keep target volumes around for migrations that didn't succeed. 

For migrations that didn't reach the completed status - not for any failed migration.

> However, that timestamp thing doesn't make sense to me. We have no idea what the state of the target PVC is as long as we don't see the migration succeed. It could very well be empty or partially populated. What we can do is keep it around and clean it up next time we migrate the VM? That way it's there for recovery purposes in case the highly unlikely scenario above occurs.

We are only speaking about the scenario when the VM is on the target and is writing to the target PVC when the source is crashed/unavailable. There should be something on the target as the migration has actually been completed. (only the source doesn't know about it and didn't update all other components)

I feel that there is a disconnect, let's take it offline when I'm back.

Reply:
@alicefr This issue and https://github.com/kubevirt/kubevirt/issues/13112 may affect the already implemented volume migration.